topologies/multi.go

```diff
@@ -80,75 +94,85 @@ func (t *MultiTopology) GenerateBGPPeers(routeReflectors []corev1.Node, nodes ma
 			},
 		}
 	}
-	selector := fmt.Sprintf("has(%s)", t.NodeLabelKey)
-	rrConfig.Spec = calicoApi.BGPPeerSpec{
-		NodeSelector: selector,
-		PeerSelector: selector,
-	}
-
-	bgpPeerConfigs = append(bgpPeerConfigs, *rrConfig)
+	if rrConfig.Spec.NodeSelector != selector || rrConfig.Spec.PeerSelector != selector {
+		rrConfig.Spec = calicoApi.BGPPeerSpec{
+			NodeSelector: selector,
+			PeerSelector: selector,
+		}
 
-	peers := int(math.Min(float64(len(routeReflectors)), 3))
+		toRefresh = append(toRefresh, *rrConfig)
+	}
+	toKeep[rrConfig.Name] = true
 
-	rrIndex := -1
-	rrIndexPerZone := map[string]int{}
+	rrsPerNode := int(math.Min(float64(len(routeReflectors)), 3))
+	isMultiZone := t.IsMultiZone(nodes)
 
+	rand.Seed(time.Now().UnixNano())
 	for n := range nodes {
 		if t.IsRouteReflector(string(n.GetUID()), n.GetLabels()) {
 			continue
 		}
 
-		routeReflectorsForNode := []corev1.Node{}
-
-		if t.Config.ZoneLabel != "" {
-			nodeZone := n.GetLabels()[t.Config.ZoneLabel]
-			rrsSameZone := []corev1.Node{}
-
-			for i, rr := range routeReflectors {
-				rrZone := rr.GetLabels()[t.Config.ZoneLabel]
-				if nodeZone == rrZone {
-					if _, ok := rrIndexPerZone[nodeZone]; !ok {
-						rrIndexPerZone[nodeZone] = -1
-					}
-
-					rrsSameZone = append(rrsSameZone, routeReflectors[i])
+		routeReflectorsForNode := t.GetRRsofNode(nodes, existingPeers, n)
+		log.Debugf("Found %v RRs for Node:%s", len(routeReflectorsForNode), n.Name)
+
+		// Select the first two RRs from different zones in Multi Zone clusters
+		// These aren't necessarily from the same cluster as the Node
+		if isMultiZone {
+
+			var firstRR *corev1.Node
+			if len(routeReflectorsForNode) == 0 {
+				firstRR = &routeReflectors[rand.Intn(len(routeReflectors))]
```

Comment:
This isn't deterministic :(

Reply:
"It only selects the first RR randomly" so not deterministic at the end of the day :(

Reply:
At that stage in the RR selection it doesn't have to be deterministic as it's uses the Node's RR state from the BGPPeers configureations.